src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+                }
 
-        Return:
-            return idx;
+                idx = -1; // all characters allowed
+
+            Return:
+                return idx;
+            }
         }
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
-            int idx;
-
-            // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
-            // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
-            if (encoder != null && !value.IsEmpty)
+            fixed (char* ptr = value)
             {
-                fixed (char* ptr = value)
+                int idx = 0;
+
+                // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
+                // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
+                if (encoder != null && !value.IsEmpty)
                 {
                     idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
+                    goto Return;
                 }
-                goto Return;
-            }
 
-            for (idx = 0; idx < value.Length; idx++)
-            {
-                if (NeedsEscaping(value[idx]))
+#if BUILDING_INBOX_LIBRARY
+                if (Sse2.IsSupported)
                 {
-                    goto Return;
+                    short* startingAddress = (short*)ptr;
+                    while (value.Length - 8 >= idx)
+                    {
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + value.Length - 8));
+
+                        // Load the next 8 characters.
+                        Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        // Check if any of the 8 characters need to be escaped.
+                        Vector128<short> mask = CreateEscapingMask(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 8 characters needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one character that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 8 characters.
+                            idx += BitOperations.TrailingZeroCount(index) >> 1;
+                            goto Return;
+                        }
+                        idx += 8;
+                        startingAddress += 8;
+                    }
+
+                    // Process the remaining characters.
+                    Debug.Assert(value.Length - idx < 8);
+                }
+#endif
+
+                for (; idx < value.Length; idx++)
```

Comment:
You could handle trailing elements and even the `length < 8` case using hwintrinsics as well.

Namely if you do `ptr -= (16 - remaining)` and then mask out the first `(16 - remaining)` elements, you can safely do a single iteration of the above vectorized code; and avoid the trailing logic here

Reply:
I tried that approach but that tend to slow things down quite a bit for length < 8.
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee#file-test_stringescaping_comparison-cs-L335-L345

For example, processing 1 character string went from 3 ns to 10 ns.

Reply:
> mask out the first (16 - remaining) elements

I think it's not needed to mask them out, so you could save that effort. Just keep them in. The check is idempotent, and it's already proven that these don't need to be escaped, so another check will result in the same.

Just post-process the correct index.

Reply:
I think the masking out would be necessary if the same loop was used for `length < 8` (and not just leftovers after `length % 8`). You wouldn't want the data that happens to be there in memory before the data (outside the string passed-in) to affect the results of escaping.

Reply:
Maybe we talk about differnt things now.

@tannergooding and I mean that the remainder of the vectorized code-path can also be processed in a vectorized way. Just feed a vector beginning from the end through the vectorized code-path above. There is no masking out needed.

Trivial example:
Let's assume length = 10 chars.
First vector has elements `0..8`, they are checked, nothing needs to be escaped. 2 elements remain.
Final vector has elements `2..10`, where six elements (`2..8`) are already checked, but we don't care, as this is cheaper than masking out, and the check is idempotent, meaning for elements `2..8` we get the same result (= nothing needs to be escaped, otherwise we wouldn't get here) as in the theck for `0..8`. So only the portion `9..10` (= the remainder) can give a different result.

This should be faster than processing the remainder with masking out or with the sequential loop (although maybe not for this example with 2 remaining elements, but for larger remainders or on the byte-version, where 16 elements are checked at one, so larger remainders can occur).

Clear what I mean?

Reply:
Yes, I understand what you mean. We could certainly do that for `length >= 8` and that's partially what I tried here. 
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee#file-test_stringescaping_comparison-cs-L312-L385

It's a good idea to avoid the masking step for the processing the trailing bits/remainder and that can make checking the remainder faster, but there are diminishing returns here and I am less concerned about the remainder for `length >= 8` case. The `NeedsEscaping_BackFill` approach is too slow for input `length < 8` which is why I didn't go down that route and your suggestions don't apply for that case to make that faster. I didn't want the logic be subtly different for `length < 8` and processing the trailing `length % 8` characters, in an effort to avoid code duplication and increasing the code complexity too much (especially since we have if/def and Sse.IsSupported blocks already that have to process sequentially).

If you have suggestions where I can process both `length < 8` and `length % 8` remaining characters in the same way, then I can explore that :)

Still, I will play around with the suggestion for processing the remainder. I don't know if adding the "remainder" logic in the main 8-character loop would be worth it though.

Reply:
So, I tried this, and it is slower when `inputLength % 8 <= 4` (that's the break-even point where the intrinsics is faster than iterating sequentially - so if the string is between 8 and 12 characters, sequential is better, but for 12 to 16, intrinsics is better).

```C#
public static unsafe int NeedsEscaping_BackFill_New(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
{
    fixed (char* ptr = value)
    {
        int idx = 0;

        if (encoder != null && !value.IsEmpty)
        {
            idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
            goto Return;
        }

        if (value.Length < 8)
        {
            for (; idx < value.Length; idx++)
            {
                if (NeedsEscaping(*(ptr + idx)))
                {
                    goto Return;
                }
            }
        }
        else
        {
            short* startingAddress = (short*)ptr;
            while (value.Length - 8 >= idx)
            {
                if (NextEightCharactersNeedEscaping(startingAddress, ref idx))
                {
                    goto Return;
                }
                idx += 8;
                startingAddress += 8;
            }

            int remainder = value.Length - idx;
            if (remainder > 0)
            {
                if (NextEightCharactersNeedEscaping(startingAddress, ref idx, 8 - remainder))
                {
                    goto Return;
                }
            }
        }

        idx = -1; // all characters allowed

    Return:
        return idx;
    }
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static bool NextEightCharactersNeedEscaping(short* startingAddress, ref int idx, int backFillCount = 0)
{
    bool result = true;
    Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress - backFillCount);

    Vector128<short> mask = CreateEscapingMask(sourceValue);

    int index = Sse2.MoveMask(mask.AsByte());
    if (index != 0)
    {
        idx += (BitOperations.TrailingZeroCount(index) >> 1) - backFillCount;
        goto Return;
    }
    result = false;
Return:
    return result;
}
```

``` ini

BenchmarkDotNet=v0.11.5.1159-nightly, OS=Windows 10.0.18362
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha1-014834
  [Host]     : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT
  Job-QDEVCA : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250.0000 ms  MaxIterationCount=20  
MinIterationCount=15  WarmupCount=1  

```
|                    Method | DataLength | NegativeIndex |       Mean |     Error |    StdDev |     Median |        Min |        Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------- |-------------- |-----------:|----------:|----------:|-----------:|-----------:|-----------:|------:|--------:|------:|------:|------:|----------:|
|      **NeedsEscapingCurrent** |          **1** |            **-1** |   **2.691 ns** | **0.1272 ns** | **0.1465 ns** |   **2.654 ns** |   **2.529 ns** |   **3.023 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          1 |            -1 |   5.254 ns | 0.2557 ns | 0.2944 ns |   5.163 ns |   4.902 ns |   5.865 ns |  1.96 |    0.10 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          1 |            -1 |   5.507 ns | 0.1617 ns | 0.1863 ns |   5.468 ns |   5.267 ns |   5.895 ns |  2.05 |    0.15 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |          **2** |            **-1** |   **3.828 ns** | **0.1258 ns** | **0.1448 ns** |   **3.832 ns** |   **3.586 ns** |   **4.043 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          2 |            -1 |   5.901 ns | 0.1807 ns | 0.2008 ns |   5.882 ns |   5.466 ns |   6.273 ns |  1.54 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          2 |            -1 |   6.445 ns | 0.2043 ns | 0.2353 ns |   6.395 ns |   6.093 ns |   6.841 ns |  1.69 |    0.09 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |          **4** |            **-1** |   **6.112 ns** | **0.2043 ns** | **0.2353 ns** |   **6.064 ns** |   **5.835 ns** |   **6.622 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          4 |            -1 |   7.649 ns | 0.1833 ns | 0.1961 ns |   7.623 ns |   7.404 ns |   8.099 ns |  1.25 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          4 |            -1 |   8.718 ns | 0.2017 ns | 0.1981 ns |   8.726 ns |   8.202 ns |   8.987 ns |  1.43 |    0.06 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |          **7** |            **-1** |   **9.393 ns** | **0.2183 ns** | **0.2042 ns** |   **9.380 ns** |   **8.935 ns** |   **9.749 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          7 |            -1 |  11.000 ns | 0.5758 ns | 0.6631 ns |  10.718 ns |  10.247 ns |  12.539 ns |  1.14 |    0.04 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          7 |            -1 |  11.972 ns | 0.2565 ns | 0.2400 ns |  12.032 ns |  11.479 ns |  12.338 ns |  1.28 |    0.04 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |          **8** |            **-1** |  **10.450 ns** | **0.2168 ns** | **0.2028 ns** |  **10.426 ns** |  **10.166 ns** |  **10.863 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          8 |            -1 |  10.929 ns | 0.5047 ns | 0.5813 ns |  11.009 ns |   9.567 ns |  12.101 ns |  1.05 |    0.04 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          8 |            -1 |  11.565 ns | 0.5790 ns | 0.6435 ns |  11.454 ns |  10.592 ns |  13.197 ns |  1.08 |    0.04 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |          **9** |            **-1** |  **11.475 ns** | **0.3124 ns** | **0.3473 ns** |  **11.408 ns** |  **10.934 ns** |  **12.328 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |          9 |            -1 |   9.772 ns | 0.3826 ns | 0.4252 ns |   9.733 ns |   9.058 ns |  10.909 ns |  0.85 |    0.04 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |          9 |            -1 |  16.270 ns | 1.2363 ns | 1.4237 ns |  15.821 ns |  14.476 ns |  18.352 ns |  1.42 |    0.12 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **10** |            **-1** |  **12.802 ns** | **0.3269 ns** | **0.3357 ns** |  **12.853 ns** |  **12.274 ns** |  **13.741 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         10 |            -1 |  11.443 ns | 0.7532 ns | 0.8673 ns |  11.784 ns |  10.070 ns |  12.654 ns |  0.91 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         10 |            -1 |  14.779 ns | 0.4825 ns | 0.5557 ns |  14.732 ns |  13.998 ns |  15.937 ns |  1.15 |    0.05 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **11** |            **-1** |  **14.341 ns** | **0.6091 ns** | **0.6771 ns** |  **14.153 ns** |  **13.406 ns** |  **15.962 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         11 |            -1 |  13.819 ns | 0.9048 ns | 1.0057 ns |  13.922 ns |  12.053 ns |  15.625 ns |  0.97 |    0.08 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         11 |            -1 |  15.823 ns | 0.7545 ns | 0.8689 ns |  15.679 ns |  14.482 ns |  17.757 ns |  1.10 |    0.08 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **12** |            **-1** |  **16.141 ns** | **0.7545 ns** | **0.8386 ns** |  **16.068 ns** |  **14.863 ns** |  **18.070 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         12 |            -1 |  13.761 ns | 0.8433 ns | 0.9711 ns |  13.533 ns |  12.539 ns |  16.092 ns |  0.85 |    0.07 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         12 |            -1 |  15.607 ns | 0.6422 ns | 0.7396 ns |  15.459 ns |  14.540 ns |  17.284 ns |  0.97 |    0.06 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **13** |            **-1** |  **16.830 ns** | **0.5668 ns** | **0.6527 ns** |  **16.802 ns** |  **15.870 ns** |  **17.968 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         13 |            -1 |  15.482 ns | 1.0933 ns | 1.1698 ns |  15.647 ns |  13.727 ns |  18.171 ns |  0.92 |    0.09 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         13 |            -1 |  15.220 ns | 0.5194 ns | 0.5982 ns |  15.183 ns |  14.068 ns |  16.489 ns |  0.90 |    0.03 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **14** |            **-1** |  **17.295 ns** | **0.3962 ns** | **0.4068 ns** |  **17.286 ns** |  **16.530 ns** |  **18.282 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         14 |            -1 |  16.075 ns | 1.0214 ns | 1.1762 ns |  16.086 ns |  14.443 ns |  18.692 ns |  0.94 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         14 |            -1 |  15.367 ns | 0.8081 ns | 0.8982 ns |  15.374 ns |  13.909 ns |  17.068 ns |  0.89 |    0.05 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **15** |            **-1** |  **19.173 ns** | **0.8490 ns** | **0.9084 ns** |  **18.867 ns** |  **17.936 ns** |  **20.772 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         15 |            -1 |  15.834 ns | 0.3529 ns | 0.3922 ns |  15.813 ns |  15.059 ns |  16.522 ns |  0.83 |    0.03 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         15 |            -1 |  14.575 ns | 0.5166 ns | 0.5949 ns |  14.416 ns |  13.751 ns |  16.156 ns |  0.76 |    0.04 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **16** |            **-1** |  **19.666 ns** | **0.6339 ns** | **0.7300 ns** |  **19.641 ns** |  **18.566 ns** |  **21.430 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         16 |            -1 |  10.682 ns | 0.3242 ns | 0.3604 ns |  10.527 ns |  10.234 ns |  11.514 ns |  0.54 |    0.02 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         16 |            -1 |  11.398 ns | 0.3076 ns | 0.3419 ns |  11.435 ns |  10.806 ns |  12.170 ns |  0.58 |    0.02 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **17** |            **-1** |  **20.502 ns** | **0.5045 ns** | **0.5809 ns** |  **20.342 ns** |  **19.643 ns** |  **21.622 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         17 |            -1 |  13.003 ns | 0.8819 ns | 1.0156 ns |  13.332 ns |  11.365 ns |  14.558 ns |  0.63 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         17 |            -1 |  18.021 ns | 1.1066 ns | 1.0868 ns |  18.265 ns |  16.467 ns |  19.709 ns |  0.88 |    0.07 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **23** |            **-1** |  **30.704 ns** | **2.1582 ns** | **2.4854 ns** |  **30.470 ns** |  **26.943 ns** |  **35.727 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         23 |            -1 |  18.486 ns | 1.1722 ns | 1.3029 ns |  18.591 ns |  16.955 ns |  21.408 ns |  0.60 |    0.06 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         23 |            -1 |  16.181 ns | 0.3261 ns | 0.2546 ns |  16.278 ns |  15.710 ns |  16.424 ns |  0.52 |    0.04 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **24** |            **-1** |  **28.256 ns** | **0.7450 ns** | **0.8580 ns** |  **28.253 ns** |  **27.107 ns** |  **29.760 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         24 |            -1 |  13.871 ns | 0.4326 ns | 0.4981 ns |  13.958 ns |  12.749 ns |  14.661 ns |  0.49 |    0.01 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         24 |            -1 |  14.986 ns | 0.6737 ns | 0.7488 ns |  14.837 ns |  13.582 ns |  16.694 ns |  0.53 |    0.03 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |         **25** |            **-1** |  **29.552 ns** | **1.0059 ns** | **1.1584 ns** |  **29.398 ns** |  **28.229 ns** |  **32.342 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |         25 |            -1 |  18.139 ns | 2.1858 ns | 2.5171 ns |  19.485 ns |  13.794 ns |  20.314 ns |  0.61 |    0.08 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |         25 |            -1 |  20.058 ns | 0.4126 ns | 0.4415 ns |  20.079 ns |  19.283 ns |  20.993 ns |  0.68 |    0.03 |     - |     - |     - |         - |
|                           |            |               |            |           |           |            |            |            |       |         |       |       |       |           |
|      **NeedsEscapingCurrent** |        **100** |            **-1** | **118.124 ns** | **2.8988 ns** | **3.2220 ns** | **117.936 ns** | **113.771 ns** | **124.588 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|     NeedsEscapingNewFixed |        100 |            -1 |  35.771 ns | 0.9760 ns | 1.1240 ns |  35.687 ns |  34.410 ns |  38.375 ns |  0.30 |    0.01 |     - |     - |     - |         - |
| NeedsEscapingBackFill_New |        100 |            -1 |  41.861 ns | 1.0144 ns | 1.1275 ns |  41.690 ns |  40.308 ns |  44.438 ns |  0.35 |    0.01 |     - |     - |     - |         - |

Reply:
> I don't think it's worth it.

After reading your results I think it isn't worth it, and stick with the simpler code. 
Thanks for sharing the results.

But for the byte-case this could be more interesting, as there are `% 16` elements to check.

--- 

To exxegerate it: "4 (that's the break-even point ...)" special-case the remainder and take the vectorized or sequential path :wink:

Reply:
> To exxegerate it: "4 (that's the break-even point ...)" special-case the remainder and take the vectorized or sequential path :wink:

I considered that, but if the first character to escape is in the first 4, that too is slower when doing intrinsics (rather than sequential), since the sequential one bails early. So, I would have to special case the special case :)